test/extended/machine_config/pinnedimages.go

```diff
@@ -0,0 +1,219 @@
+		oc                               = exutil.NewCLIWithoutNamespace("machine-config")
+	)
+	// Ensure each test pins a separate image, since we are not deleting them after each
+
+	g.It("All Nodes in a custom Pool should have the PinnedImages even after Garbage Collection [apigroup:machineconfiguration.openshift.io]", func() {
+		// Create custom MCP
+		err := oc.Run("apply").Args("-f", customMCP).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer deleteMCP(oc, "custom")
+
+		// Add node to pool
+		nodes, err := oc.KubeClient().CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: labels.SelectorFromSet(labels.Set{"node-role.kubernetes.io/worker": ""}).String()})
+		o.Expect(err).NotTo(o.HaveOccurred(), "Get all nodes")
+		err = oc.AsAdmin().Run("label").Args("node", nodes.Items[0].Name, "node-role.kubernetes.io/custom=").Execute()
+		defer unlabelNode(oc, nodes.Items[0].Name)
+
+		// Apply KC to Pool
+		err = oc.Run("apply").Args("-f", customGcKC).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer deleteKC(oc, "custom-gc-config")
+
+		SimplePISTest(oc, customGCMCPpinnedImageSetFixture, true)
+	})
+
+	g.It("All Nodes in a Custom Pool should have the PinnedImages in PIS [apigroup:machineconfiguration.openshift.io]", func() {
+		// Create custom MCP
+		err := oc.Run("apply").Args("-f", customMCP).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer deleteMCP(oc, "custom")
+
+		// Add node to pool
+		nodes, err := oc.KubeClient().CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: labels.SelectorFromSet(labels.Set{"node-role.kubernetes.io/worker": ""}).String()})
+		o.Expect(err).NotTo(o.HaveOccurred(), "Get all nodes")
+		err = oc.AsAdmin().Run("label").Args("node", nodes.Items[0].Name, "node-role.kubernetes.io/custom=").Execute()
+		defer unlabelNode(oc, nodes.Items[0].Name)
+
+		SimplePISTest(oc, customMCPpinnedImageSetFixture, true)
+	})
+
+	g.It("All Nodes in a standard Pool should have the PinnedImages PIS [apigroup:machineconfiguration.openshift.io]", func() {
+		SimplePISTest(oc, pinnedImageSetFixture, true)
+	})
+
+	g.It("Invalid PIS leads to degraded MCN in a standard Pool [apigroup:machineconfiguration.openshift.io]", func() {
+		SimplePISTest(oc, invalidPinnedImageSetFixture, false)
+	})
+
+	g.It("Invalid PIS leads to degraded MCN in a custom Pool [apigroup:machineconfiguration.openshift.io]", func() {
+		// Create custom MCP
+		err := oc.Run("apply").Args("-f", customMCP).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer deleteMCP(oc, "custom")
+
+		// Add node to pool
+		nodes, err := oc.KubeClient().CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: labels.SelectorFromSet(labels.Set{"node-role.kubernetes.io/worker": ""}).String()})
+		o.Expect(err).NotTo(o.HaveOccurred(), "Get all nodes")
+		err = oc.AsAdmin().Run("label").Args("node", nodes.Items[0].Name, "node-role.kubernetes.io/custom=").Execute()
```

Comment:
nit: please handle this error and do we need to check length before we access Name?

Reply:
Yes updated